File: 4-1.py
#!/usr/bin/env python

from functools import reduce

assignments = [row.strip().split(",") for row in open('4.input').readlines()]
assignments = map(lambda a: list(map(lambda b: b.split("-"), a)), assignments)


def overlaps(a, b):
    if (int(a[0]) <= int(b[0]) and int(a[1]) >= int(b[1])) or \
       (int(b[0]) <= int(a[0]) and int(b[1]) >= int(a[1])):
        return 1
    else:
        return 0


print(reduce(lambda a, b: a + b, map(lambda a: overlaps(a[0], a[1]), assignments)))

File: 4-1.rb
#!/usr/bin/env ruby

assignments = File.read('4.input').lines.map(&:strip)
                .map { |s| s.split(',').map { |a| a.split('-').map(&:to_i) } }
overlap = assignments.map do |a|
  (a[0][0] <= a[1][0] and a[0][1] >= a[1][1]) or
    (a[1][0] <= a[0][0] and a[1][1] >= a[0][1])
end
print overlap.count { |a| a }, "\n"
